Add/guard installation flow

gcsecsey opened the pull request:
#### Changes proposed in this Pull Request

- [x] prevents the installation if the URL is visited without clicking the Install button on the plugins page
- [x] if plugin `A` was installed previously, it still prevents the installation of plugin `B` from the URL directly

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

### Marketplace flow

#### Prevents direct access
1.  choose a site with any plan
2. visit `/marketplace/<pluginname>/install/<siteURL>`
3. verify that the this error message is shown:
![image](https://user-images.githubusercontent.com/11555574/143490217-78c1a504-725f-4028-bfc6-f79af694e67e.png)

#### Allows install after `Install` button click
1.  choose a site with any plan
2. visit `/plugins/<pluginname>/<siteURL>`
3. click the `Install` or `Upgrade and Install` button
4. perform the upgrade if needed
5. verify that the install executes normally

#### Throws incorrect plan error message correctly
1.  choose a site with a free plan
2. visit `/plugins/<pluginname>/<siteURL>`
3. click the `Upgrade and Install` button
4. from the checkout page, click the X button in the upper left corner
5. verify that the this error message is shown:
![image](https://user-images.githubusercontent.com/11555574/143490112-2e5e5606-c216-4671-bb6f-492ac5347507.png)

#### Prevents direct access after a successful plugin install
1.  choose a site with any plan
2. visit `/plugins/<pluginname>/<siteURL>`
3. click the `Install` or `Upgrade and Install` button
4. perform the upgrade if needed
5. verify that the install executes normally
6. visit `/marketplace/<secondpluginname>/install/<siteURL>`
7. verify that the this error message is shown:
![image](https://user-images.githubusercontent.com/11555574/143490217-78c1a504-725f-4028-bfc6-f79af694e67e.png)

### Plugin upload flow

#### Prevents direct access
1.  choose a site with any plan
2. visit `/marketplace/install/<siteURL>`
3. verify that the this error message is shown:
![image](https://user-images.githubusercontent.com/11555574/144214792-ba400c70-237c-40c3-b688-ecbd4854f32a.png)

#### Allows install after plugin upload
1.  choose a site with any plan
2. visit `/plugins/upload/<siteURL>`
3. upload a plugin
4. verify that the install executes normally

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes #58297

gcsecsey committed:
[WIP] add state var when installing plugins from the UI

gcsecsey committed:
[WIP] prevent plugin installation if the URL is visited directly

Still need to clean up local state (nested ternary) and translation

gcsecsey committed:
[WIP] update error handling

gcsecsey committed:
[WIP] fail early when loading the URL directly

gcsecsey commented:
When displaying the error pages, there's a `margin-top: 350px` rule applied on the container:
![image](https://user-images.githubusercontent.com/11555574/143436129-cd97d83c-607e-4443-9db5-489b972fd825.png)

@cpapazoglou @WBerredo any ideas why it's needed or if we can remove it perhaps?

cpapazoglou commented:
> When displaying the error pages, there's a `margin-top: 350px` rule applied on the container: ![image](https://user-images.githubusercontent.com/11555574/143436129-cd97d83c-607e-4443-9db5-489b972fd825.png)
> 
> @cpapazoglou @WBerredo any ideas why it's needed or if we can remove it perhaps?

I suppose it was developed that it appears vertically centered? Probably we can make it better by utilizing `flex`?

![SS 2021-11-25 at 13 55 28](https://user-images.githubusercontent.com/12430020/143437514-4609eab8-6338-4741-ac78-bf99685a1bfe.png)

gcsecsey committed:
center error message vertically

gcsecsey committed:
update error handling

gcsecsey committed:
unset state var when plugin installation is complete

gcsecsey committed:
fix error handling

gcsecsey commented:
As we're handling multiple types of errors now in this component, I've tried consolidating the error handling with the help of an enum, but ran into some race condition issues down the line.

I see we're checking some conditions in an async way here: https://github.com/Automattic/wp-calypso/blob/0d38fdc2df4bfeb26135270a69f6d6e0e83cc8c5/client/my-sites/marketplace/pages/marketplace-plugin-upload-status/index.tsx#L98

@WBerredo could you help me with why the callback is needed here? Do we have some piece of state that changes on page load? Thanks 🙂

WBerredo commented:
This `useEffect` will run at every render (we don't pass the properties array) and check there is an unsupported plan, if there is, it will check again in 2 seconds and if the problem persists, it shows an error.
https://github.com/Automattic/wp-calypso/blob/0d38fdc2df4bfeb26135270a69f6d6e0e83cc8c5/client/my-sites/marketplace/pages/marketplace-plugin-upload-status/index.tsx#L96-L102


> why the callback is needed here?

On the **upgrade and install** flow, the plan we receive on the first render is the old/low tier plan, but this plan gets updated on the subsequent renders. That's why we wait 2 seconds before showing the error, and we consider if it didn't get updated in 2 seconds this is the current plan of the user.

---

Does that makes sense? There is anything that still needs explaining?

gcsecsey commented:
@WBerredo I thought that we can't use the data that we have on the initial render, but wasn't sure, thanks for clarifying! 🎉

WBerredo reviewed (commented):
I was able to test the general path, but I would like to point two things:

* I didn't get the "upgrade to business plan" error, when following the steps I got the "direct access" error.
* Shouldn't we store which plugin triggered the install and check it on the state? Otherwise, I would be able to trigger an install and access the direct URL with another plugin installation if done at the same time.

gcsecsey commented:
Thanks for testing this @WBerredo, it'd be a great idea to add the plugin to the state! I'll check the issue with the errors and add this additional check. 🙂

gcsecsey committed:
verify plugins slug and domain name when installing plugin from URL

gcsecsey committed:
fix plan upgrade error message

gcsecsey committed:
Merge branch 'trunk' into add/guard-installation-flow

gcsecsey committed:
delete unnecessary whitespace

cpapazoglou commented on client/my-sites/plugins/plugin-details.jsx:
```diff
@@ -436,6 +441,15 @@ function onClickInstallPlugin( { dispatch, selectedSite, slug, upgradeAndInstall
 		} )
 	);
 
+	dispatch( productToBeInstalled( null, slug, selectedSite.slug ) );
+
+	dispatch(
+		pluginInstallationStateChange(
+			MARKETPLACE_ASYNC_PROCESS_STATUS.IN_PROGRESS,
+			'preauthorize plugin installation URL'
+		)
+	);
+
```
We will need to add similar events to the Plugin Upload flow `https://wordpress.com/plugins/upload/[DOMAIN]` which is now returning an error. 

https://github.com/Automattic/wp-calypso/blob/08961f239b9c353bb0b44983d9f7b014db038133/client/my-sites/plugins/plugin-upload/index.jsx#L60-L62

gcsecsey replied:
Thanks @cpapazoglou, that's a very good point!

gcsecsey committed:
guard plugin upload flow too

gcsecsey committed:
add new error message to plugin upload flow

gcsecsey commented:
> Shouldn't we store which plugin triggered the install and check it on the state? Otherwise, I would be able to trigger an install and access the direct URL with another plugin installation if done at the same time.

>We will need to add similar events to the Plugin Upload flow https://wordpress.com/plugins/upload/[DOMAIN] which is now returning an error.

I've added these changes and updated the test instructions with steps for the 'plugin upload' flow. When trying to visit the install page directly with the plugin upload URL, we're now showing a flow-specific error message that helps the user navigate back to the plugin upload page.

> I didn't get the "upgrade to business plan" error, when following the steps I got the "direct access" error.

This was a small mistake on my side, I've accidentally copied over the other error message. 😄 I've fixed it now, it should work as per the test instructions. 👍

WBerredo reviewed (changes requested):
I am seeing an error on the **upgrade and install flow**, if I complete the purchase or not I am always getting the "This URL should not be accessed directly. Please click the Install button on the plugin page." error.

I'm not sure if this is related to the environment it is created on the PR. But I think this worth some research.

gcsecsey committed:
verify only `primaryDomain` and `productSlug` when installing plugins

`pluginInstallationStatus` could be overwritten by other actions

gcsecsey committed:
Merge branch 'trunk' into add/guard-installation-flow

gcsecsey commented:
> I am seeing an error on the upgrade and install flow, if I complete the purchase or not I am always getting the "This URL should not be accessed directly. Please click the Install button on the plugin page." error.

I've looked into this, and found that previously, the `pluginInstallationStatus` state var was written from both actions that were fired, resulting in a race condition:
https://github.com/Automattic/wp-calypso/blob/add/guard-installation-flow/client/state/marketplace/purchase-flow/reducer.ts#L38

I've updated the installation check to still validate the `primaryDomain` and `productSlug` variables, and to check that the plugin installation hasn't been completed yet:
https://github.com/Automattic/wp-calypso/pull/58493/files#diff-b9d1e3ec36a8799dd188abca2d80e61b898eb69540f0fdb44742afd69ccc8e95R82

https://github.com/Automattic/wp-calypso/pull/58493/files#diff-b9d1e3ec36a8799dd188abca2d80e61b898eb69540f0fdb44742afd69ccc8e95R87

gcsecsey committed:
evaluate error causes in priority order

gcsecsey commented:
With @WBerredo we've discovered a new edge case.

When the user navigates away from the checkout page by changing the URL in the browser, the IndexedDB state is reset to the `defaultState` value:
https://d.pr/v/rDDfH9

However, when the user navigates away from the checkout page by clicking the X icon in the upper left corner of the checkout page, the IndexedDB state is persisted as expected:
https://d.pr/v/tqbGqp

It seems that by navigating to a new URL, the Redux init clears the persisted state instead of loading the values from IndexedDB, and hence breaks the purchase flow. When navigating through client routes, the `defaultState` isn't loaded again and the feature works as expected.

We've tried removing the `defaultState` so that values in IndexedDB can be read back on a later visit. However, this results in an error, the slice reducer can't be initiated without a default state:
![image](https://user-images.githubusercontent.com/11555574/144653769-8df39c2e-939a-4c07-ba27-8ee01bc1dd92.png)

WBerredo reviewed (approved):
Tested the last version and it is all working good 🚀

PS: The only edge case I found is mentioned at this PR but I don't think we can do anything about it.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7024500

Thank you @gcsecsey for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7024500).
